internal/backend/workloadproxy/reconciler.go

```diff
@@ -6,225 +6,374 @@
+		rec.mu.Lock()
+		defer rec.mu.Unlock()
+
+		clusterID, rawPort, splitErr := net.SplitHostPort(addr)
+		if splitErr != nil {
+			return nil, fmt.Errorf("invalid proxy dst address: %w", splitErr)
+		}
+
+		p := port(rawPort)
+
+		ad := rec.aliasToCluster.ClusterData(clusterID)
+		if ad == nil {
+			return nil, fmt.Errorf("cluster %q not found", clusterID)
+		}
+
+		got := rec.activeProbes.Get(clusterID)
+		if ad.inUsePort == "" {
+			if got == nil {
+				logger := rec.logger.With(
+					zap.String("cluster", clusterID),
+					zap.Strings("upstreams", ad.hosts),
+					zap.Stringer("probe_port", p),
+				)
+
+				l, lErr := upstream.NewListWithCmp(asRemotes(ad.hosts, p, logger), (*remote).Equal)
+				if lErr != nil {
+					return nil, fmt.Errorf("failed to create upstream list for cluster %q: %w", clusterID, lErr)
+				}
+
+				got = rec.activeProbes.Set(clusterID, l, ttlcache.DefaultTTL)
+			}
+
+			err := rec.aliasToCluster.SetActivePort(clusterID, p)
+			if err != nil {
+				return nil, fmt.Errorf("failed to set active lb port %q: %w", p, err)
+			}
+		}
 
-	proxy := httputil.NewSingleHostReverseProxy(targetURL)
-	proxy.Transport = &http.Transport{
-		DialContext: registry.connProvider.DialContext,
+		switch rem, err := got.Value().Pick(); {
+		case err != nil:
+			return nil, fmt.Errorf("failed to pick upstream for cluster %q: %w", clusterID, err)
+		case rem == nil:
+			return nil, fmt.Errorf("failed to pick upstream for cluster %q", clusterID)
+		default:
+			return d.DialContext(ctx, network, net.JoinHostPort(rem.Addr, p.String()))
+		}
 	}
 
-	return proxy, clusterID, nil
+	rec.transport = wrapWithMetrics(transport, rec)
 }
 
-func (registry *Reconciler) removeLB(cluster resource.ID, alias string) {
-	registry.logger.Log(registry.logLevel, "remove LB", zap.String("cluster", cluster), zap.String("alias", alias))
+// Describe implements the [prometheus.Collector] interface.
+func (rec *Reconciler) Describe(ch chan<- *prometheus.Desc) { prometheus.DescribeByCollect(rec, ch) }
 
-	aliasToLB := registry.clusterToAliasToLBStatus[cluster]
-	lbSts := aliasToLB[alias]
+// Collect implements the [prometheus.Collector] interface.
+func (rec *Reconciler) Collect(mcs chan<- prometheus.Metric) { rec.idleConnections.Collect(mcs) }
 
-	if lbSts != nil {
-		if err := lbSts.Close(); err != nil {
-			registry.logger.Error("failed to close LB", zap.String("cluster", cluster), zap.String("alias", alias), zap.Error(err))
+func asRemotes(hosts []string, p port, l *zap.Logger) iter.Seq[*remote] {
+	return func(yield func(*remote) bool) {
+		for _, h := range hosts {
+			if !yield(&remote{Logger: l, Addr: h, AddrPort: net.JoinHostPort(h, string(p))}) {
+				return
+			}
 		}
 	}
+}
 
-	delete(aliasToLB, alias)
-	delete(registry.aliasToCluster, alias)
+// setEvictionFunc sets the eviction function for the activeProbes cache. It will shut down the active load balancer
+// probe.
+func (rec *Reconciler) setEvictionFunc() func() {
```

Comment:
would inlining this into the call site make it more readable? 

it seems it can be reduced to `res.activeProbes.OnEviction(func() { ... })` ?